### ISSUE TITLE
[WAITING ON MIROSLAV FOR PAYMENT] [HOLD for payment 2023-08-28] [$1000] IOS - New contact method - "New contact method" button abrupt animation when tap on email

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#19530]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Issue found when executing PR** https://github.com/Expensify/App/pull/19040

## Action Performed:
1. Open https://staging.new.expensify.com/
2. Tap profile icon
3, Tap "Profile"
4. Tap "Contact method"
5. Tap "New contact method"
6. Type an email
7. Tap "Add"
8. Tap on email from step 6 to confirm it

## Expected Result:
When user taps on email a smooth animation of transition from the page "Contact methods" to email confirmation page appears

## Actual Result:
When user taps on email "New contact method" button slides up and appears cut before user is transferred to email confirmation page

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.17.0

**Reproducible in staging?:** yes

**Reproducible in production?:** n/a

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://platform.applause.com/services/links/v1/external/5a5bc5fdc111765130d79b24b0362e74f114b5358cf8d0952195a66b00d1e5c7

https://platform.applause.com/services/links/v1/external/661a2aa54f61e6426e10547569cd9407700e54124c4968b7a6347d6a4390f16f

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0123c376fe9502d376</li>
        <li>Upwork Job ID: 1671611450736222208</li>
        <li>Last Price Increase: 2023-08-04</li>
<li>Automatic offers: </li>
<ul>
<li>situchan | Contributor | 26067330</li>
</ul></li>
    </ul>
</details>

**Comment by hungvu193** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
IOS - New contact method - "New contact method" button abrupt animation when tap on email
### What is the root cause of that problem?
After click the contact methods, our app is navigating to Contact Method Detail page, and the keyboard opened, this also make the current page (which is not fully unmounted) push the bottom button since it's still navigating.
### What changes do you think we should make in order to solve the problem?
We should wait for the transition ended and then rendered the content of Contact Detail Page, we can avoid this issue.
So we can add `onEntryTransitionEnd` function of `ContactMethodDetailsPage` like this:
```javascript
            <ScreenWrapper
                onEntryTransitionEnd={() => {
                    this.setState({ isTransitionEnd: true })
                }}
```
and update the content of it like this:
```javascript
 {this.state.isTransitionEnd ? <ScrollView ></ScrollView> : <FullScreenLoadingorSkeletonLoading />
```

### What alternative solutions did you explore? (Optional)
N/A

**Comment by hungvu193** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
IOS - New contact method - "New contact method" button abrupt animation when tap on email
### What is the root cause of that problem?
After click the contact methods, our app is navigating to Contact Method Detail page, and the keyboard opened, this also make the current page (which is not fully unmounted) push the bottom button since it's still navigating.
### What changes do you think we should make in order to solve the problem?
We should wait for the transition ended and then rendered the content of Contact Detail Page, we can avoid this issue.
So we can add `onEntryTransitionEnd` function inside `ContactMethodDetailsPage` like this:
```javascript
            <ScreenWrapper
                onEntryTransitionEnd={() => {
                    this.setState({ isTransitionEnd: true })
                }}
```
and update the content of it like this:
```javascript
 {this.state.isTransitionEnd ? <ScrollView ></ScrollView> : <FullScreenLoadingorSkeletonLoading />
```

### What alternative solutions did you explore? (Optional)
N/A
### Result:

https://github.com/Expensify/App/assets/16502320/69082b0f-bfbc-47f2-ba45-2ce028bc239d

**Comment by jliexpensify** [comment]
Hi @0xmiroslav - @kbecciv stated this was related to your PR https://github.com/Expensify/App/pull/19040

Could you have a look at this issue? Cheers!

**Comment by jliexpensify** [comment]
Thanks! @joelbettner do you mind weighing in if this is worth fixing?

**Comment by joelbettner** [comment]
This is definitely worth fixing, I'm a bit confused by the proposal, though. It looks like this behavior is happening in multiple places (the new contact method and the entering of the magic code). Is the proposal for fixing this in all places, or only in one place? I think we should be fixing this in all places.

**Comment by hungvu193** [comment]
My proposal only fixed it in this screen. Yeah, there're other screen with similar behavior, they are doing similar thing as this proposal, but in stead of show loading, they are waiting for transition end and focus the ref.
https://github.com/Expensify/App/blob/7930cb59313f17750fb1670e19b6fee05fc0245b/src/pages/settings/Profile/Contacts/NewContactMethodPage.js#L112-L118
https://github.com/Expensify/App/blob/7930cb59313f17750fb1670e19b6fee05fc0245b/src/pages/settings/Payments/AddPayPalMePage.js#L46

**Comment by joelbettner** [comment]
@hungvu193 @0xmiroslav is there a way we can fix this in one place for all instances or does it need to be fixed on a case by case basis?

**Comment by hungvu193** [comment]
I think there is a way to do that, that we will add a skeleton or loading view in ScreenWrapper, and add a new props called `shouldShowLoadingWhileTransitioning`. So we will show the loading until the entry transition ended.

**Comment by joelbettner** [comment]
@hungvu193 @0xmiroslav any updates here?

**Comment by jliexpensify** [comment]
Not overdue, still under review.

**Comment by jliexpensify** [comment]
@joelbettner @0xmiroslav hello, how are we going with this one? @hungvu193 no luck with eliminating the loading?

**Comment by joelbettner** [comment]
> There should be solution without loading.

> @joelbettner I couldn't find any better idea rather than show a loading before transition end :(

@0xmiroslav any chance you can help @hungvu193 with ideas that do not include loading...or do we possibly go with a solution that includes loading?

**Comment by joelbettner** [comment]
I added the External and Help Wanted labels. I think we are just waiting on proposals now...correct?

FYI, I'm going to be OOO for a good portion of next week and the week after that. So, if you need to, don't hesitate to bring in another Engineer on this issue.

**Comment by jliexpensify** [comment]
Not overdue - bump @0xmiroslav as we have a new proposal.

**Comment by jliexpensify** [comment]
bumping @0xmiroslav for a proposal review!

**Comment by ShogunFire** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

On pages where we have the magiccodeinput the transition animation can be abrupt

### What is the root cause of that problem?

The root cause of that problem is that we focus thye textinput before the animation is finished.

In MagicCodeInput we have a prop shouldDelayFocus here:
https://github.com/Expensify/App/blob/7930cb59313f17750fb1670e19b6fee05fc0245b/src/components/MagicCodeInput.js#L30

but it doesn't work for two reasons:

Here:
https://github.com/Expensify/App/blob/7930cb59313f17750fb1670e19b6fee05fc0245b/src/components/MagicCodeInput.js#L144-L148
we focus the input regardless of the shouldDelayFocus

And here: 
https://github.com/Expensify/App/blob/7930cb59313f17750fb1670e19b6fee05fc0245b/src/components/MagicCodeInput.js#L302

We pass autofocus to the textinput 

### What changes do you think we should make in order to solve the problem?

We should only focus the input immediatly when there is no shouldDelayFocus so we should add a else like this:
```
if (props.shouldDelayFocus) {
      focusTimeout = setTimeout(() => inputRefs.current[0].focus(), CONST.ANIMATED_TRANSITION);
  }
  else{
      inputRefs.current[0].focus();
  }
```
and we should not pass the autofocus to the TextInput, in other components we don't do it either because we decided to handle it ourself.

Finally if we don't want to have to pass the props shouldDelayFocus everywhere, we can put the default value to true


### What alternative solutions did you explore? (Optional)

**Comment by jliexpensify** [comment]
@trjExpensify Hello! I am OOO the rest of this week, so just assigning a buddy to take care of hiring.

Atm, we are still sourcing proposals. I'll unassign you when I get back! Cheers.

**Comment by ShogunFire** [comment]
@0xmiroslav Hello, can you check my [proposal](https://github.com/Expensify/App/issues/19530#issuecomment-1618129707) , this is how we handle this problem everywhere in the app.

**Comment by joelbettner** [comment]
Although this is fixing as a "case-by-case" basis, if we are currently using the solution proposed by @ShogunFire elsewhere in the app, then I believe we should continue to use that method to solve it for this case. cc: @0xmiroslav @trjExpensify

**Comment by jliexpensify** [comment]
I'm back, unassigning Tom! Bumping @0xmiroslav to review the above proposals.

**Comment by joelbettner** [comment]
@0xmiroslav can you elaborate about your thoughts on this proposal: https://github.com/Expensify/App/issues/19530#issuecomment-1605291482

**Comment by jliexpensify** [comment]
Bumping @0xmiroslav for Joel's comment!

**Comment by ShogunFire** [comment]
@0xmiroslav  I was on another issue where they are already  doing it: https://github.com/Expensify/App/issues/22388#issuecomment-1634814455

**Comment by situchan** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
While screen transition, if next page has auto focus input, bottom view of previous page floats above keyboard.

### What is the root cause of that problem?
`ScreenWrapper` has keyboard avoiding behavior enabled by default

### What changes do you think we should make in order to solve the problem?
I noticed that this issue was fixed in #20608 by disabling keyboard avoiding behavior, but only contact page
https://github.com/Expensify/App/blob/2722481852846884613d8795376c4877f677ef51/src/pages/settings/Profile/Contacts/ContactMethodsPage.js#L114

Still issue happens on other pages - money request, 2fa, confirm task
We can apply same fix to those pages as they don't have keyboard display.
- MoneyRequestAmountPage
- CodesPage
- NewTaskPage

**Comment by jliexpensify** [comment]
Bump @0xmiroslav to review the above comment

**Comment by jliexpensify** [comment]
Bump @0xmiroslav for proposal reviews

**Comment by joelbettner** [comment]
@0xmiroslav I am fine with @situchan's proposal, but I don't like that it isn't general. I like that @dukenv0307's proposal is trying to fix the general case, but you bring up a good point about unforseen regressions. I'm fine with either proposal and will leave it up to you to decide.

**Comment by jliexpensify** [comment]
Bump @0xmiroslav for the decision please!

**Comment by jliexpensify** [comment]
Bump @0xmiroslav for the decision please!

**Comment by jliexpensify** [comment]
@0xmiroslav - bumping you on a [decision](https://github.com/Expensify/App/issues/19530#issuecomment-1654719208)

**Comment by jliexpensify** [comment]
Payment summary:

- Bug Reporter: @situchan $250
- Contributor: @situchan $1000 + $500 (bonus)
- C+: @0xmiroslav $1000 + $500 (bonus)

Does that sound correct?

**Comment by jliexpensify** [comment]
Great, could you complete the checklist @0xmiroslav ? Thanks!

**Comment by jliexpensify** [comment]
Great thanks, once you accept @0xmiroslav I will pay out everyone and close this.

**Comment by jliexpensify** [comment]
Paid @situchan , waiting on Miroslav

**Comment by jliexpensify** [comment]
Not overdue, just waiting on Miroslav

**Comment by jliexpensify** [comment]
@0xmiroslav can you accept the offer so I can pay you and close this out? Thanks in advance!

**Comment by jliexpensify** [comment]
Are you still sorting out your account with Upworks? Is there an ETA on when you might be ready to accept payments?

**Comment by jliexpensify** [comment]
@0xmiroslav has your account been sorted out? Can I pay into it now? Cheers.

**Comment by jliexpensify** [comment]
@0xmiroslav ready for payment?

**Comment by jliexpensify** [comment]
Thanks, I'll close it to avoid others not getting GH emails. Can you please tag me in this GH when you're ready and I will finalise payment in Upworks? Thanks!

[Payment summary comment](https://github.com/Expensify/App/issues/19530#issuecomment-1694922134).
